### DISCUSSION TITLE
danfo using CDN, reading writing csv or json files (maybe a bug)

### DISCUSSION THREAD

**Discussion by idhamari** [opensource9ja/danfojs#468] [Q&A]
Dear all, I am using CDN so I am trying to do simple stuff in the client side. 

I can read csv file but nothing works after that e.g. 

this line works:
 
           var df = dfd.readCSV ("test.csv").then(df => {df.head().print()});

but this one not (no output): 

           var df = dfd.readCSV ("test.csv")
           df.print

 or this one (no output file): 

           var df = dfd.readCSV ("test.csv")
           dfd.toCSV(df, { filePath: "testOut.csv"});

**Comment by risenW** [comment]
Check this: https://danfo.jsdata.org/api-reference/input-output/danfo.to_csv#convert-dataframe-to-csv-string-and-download-file-in-client-side-lib

**Comment by idhamari** [comment]
Thanks for your quick reply, it seems to work but when I try:

*   dfd.toCSV(df, { filePath: "testOut.csv"}); without the download option it didn't work (e.g., I want to save in the working directory not in the download folder)
* using my own csv file, it didn't work. The file has a valid format and can be read and correctly using this line:
     
       var df = dfd.readCSV ("test.csv").then(df => {df.head().print()});

**Comment by idhamari** [comment]
Moreover:

I tried to read the same saved file from the example e.g.
 
       var df = dfd.readCSV ("testOut.csv").then(df => {df.head().print()});
and I get this error:

           Uncaught (in promise) Error: Table must have a consistent number of cells.

**Comment by idhamari** [comment]
Some notes: 

* It seems when saving csv, an extra empty line is written which causes the error above **(shall I report a bug?)** my system is Windows 10 with German localization. 

            Uncaught (in promise) Error: Table must have a consistent number of cells.

* The file will be saved to the download folder, and the reading will be from the working folder.
* When reading a json file an error appears:
 
        Error: ParamError: File not supported. file must be a url or an input File object
     

Here is a complete example, note after saving, copy the file manually to the working folder next to the html file:

 
                     <!DOCTYPE html>
                <html>

                    <head>
                        <meta charset="utf-8">
                        <script src="https://cdn.jsdelivr.net/npm/danfojs@1.1.0/lib/bundle.min.js"></script>
                </head>

                <body>
                    <h1 style="color:rgb(95, 55, 10)"> Data Test </h1>               
                
                                <br>  
                                <button id="btnWriteCSV" type="button"   >Write CSV</button> 
                                <br>  
                                <button id="btnReadCSV"  type="button"   >Read CSV</button> 

                                <br>  
                                <button id="btnWriteJSN" type="button"   >Write JSON</button> 
                                <br>  
                                <button id="btnReadJSN"  type="button"   >Read JSON</button> 

                                <script>        
                                    var myFileNameCSV= "testCSVOut.csv"                 
                                    var myFileNameJSN= "testCSVOut.json"                 

                                    function writeDataCSV(){
                                        console.log(" Write csv file  .............");
                                        let data = {
                                            Abs: [20.2, 30, 47.3],
                                            Count: [34, 4, 5],
                                            "country code": ["NG", "FR", "GH"]
                                        };

                                        let df = new dfd.DataFrame(data);
                                        // Note: it is saved in Downloads folder 
                                        dfd.toCSV(df, { fileName: myFileNameCSV, download: true});

                                    }
                
                                    function readDataCSV(){
                                        console.log(" Reading csv file  .............");
                                        // copy the file to the working folder  
                                        dfd.readCSV( myFileNameCSV ) 
                                        .then(df => {                        
                                        df.head().print()

                                        })
                                    }

                                    function writeDataJSN(){
                                        console.log(" Write json file  .............");
                                        let data = [
                                                        {
                                                            "Abs" : "20.2",
                                                            "Count" : "34",
                                                            "country code" : "NG",
                                                        },
                                                        {
                                                            "Abs" : "30",
                                                            "Count" : "4",
                                                            "country code" : "FR",
                                                        },
                                                        {
                                                            "Abs" : "47.3",
                                                            "Count" : "5",
                                                            "country code" : "GH",
                                                        }
                                                   ]

                                        let df = new dfd.DataFrame(data);
                                        // Note: it is saved in Downloads folder 
                                        dfd.toJSON(df, { fileName: myFileNameJSN, download: true});

                                    }
                
                                    function readDataJSN(){
                                        console.log(" Reading json file  .............");
                                        // copy the file to the working folder  
                                        dfd.readJSON( myFileNameJSN ) 
                                        .then(df => {                        
                                        df.head().print()

                                        })
                                    }
                                    const buWriteCSV = document.querySelector('#btnWriteCSV');
                                    buWriteCSV.addEventListener('click',  writeDataCSV);

                                    const buReadCSV = document.querySelector('#btnReadCSV');
                                    buReadCSV.addEventListener('click',  readDataCSV);

                                    const buWriteJSN = document.querySelector('#btnWriteJSN');
                                    buWriteJSN.addEventListener('click',  writeDataJSN);

                                    const buReadJSN = document.querySelector('#btnReadJSN');
                                    buReadJSN.addEventListener('click',  readDataJSN);
                                </script>
                </body>

                </html>
